File: Algorithms/Python/0404_Sum_of_Left_Leaves/Sum_of_Left_Leaves_Solution_1.py

# Space: O(n)
# Time: O(n)

# BFS approach

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def sumOfLeftLeaves(self, root):
        if root is None: return 0

        queue = [root]
        res = 0

        while queue:
            cur = queue.pop(0)

            if cur.left:
                if cur.left.left is None and cur.left.right is None: # verify if this is left leaf
                    res += cur.left.val
                queue.append(cur.left)

            if cur.right:
                queue.append(cur.right)
        return res

File: Algorithms/0404_Sum_of_Left_Leaves/Java/SumofLeftLeavesSolution1.java

// Space: O(1)
// Time: O(n)

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int res = 0;

    public int sumOfLeftLeaves(TreeNode root) {
        dfs(root,false);
        return res;
    }

    private void dfs(TreeNode root, boolean flag) {
        if (root == null) return;
        if (root.left == null && root.right == null) {
            res += flag ? root.val : 0;
        }
        dfs(root.left, true);
        dfs(root.right, false);
    }
}




